pkg/keycloak/realm/phaseHandler.go

```diff
@@ -0,0 +1,449 @@
+	kcClient, err := ph.getClient(kcr)
+	if err != nil {
+		return kcr, errors.Wrapf(err, "error reconciling keycloak realm: '%v'", kcr.Spec.Realm)
+	}
+
+	if err := ph.reconcileUsers(kcClient, *kcr, kcr.ObjectMeta.Namespace); err != nil {
+		return nil, err
+	}
+
+	if err := ph.reconcileClients(kcClient, *kcr, kcr.ObjectMeta.Namespace); err != nil {
+		return nil, err
+	}
+
+	if err := ph.reconcileIdentityProviders(kcClient, *kcr); err != nil {
+		return nil, err
+	}
+
+	return kcr, nil
+}
+
+func (ph *phaseHandler) reconcileUsers(kcClient keycloak.KeycloakInterface, realm v1alpha1.KeycloakRealm, ns string) error {
+
+	users, err := kcClient.ListUsers(realm.Spec.Realm)
+	if err != nil {
+		return err
+	}
+
+	kcUsers := map[string]*v1alpha1.KeycloakUser{}
+	for i := range users {
+		kcUsers[users[i].UserName] = users[i]
+	}
+
+	userPairsList := map[string]*v1alpha1.KeycloakUserPair{}
+	for i := range realm.Spec.Users {
+		user := &realm.Spec.Users[i]
+		userPairsList[user.UserName] = &v1alpha1.KeycloakUserPair{
+			SpecUser: user,
+			KcUser:   kcUsers[user.UserName],
+		}
+	}
+
+	for i := range userPairsList {
+		err := ph.reconcileUser(userPairsList[i].KcUser, userPairsList[i].SpecUser, realm.Spec.Realm, kcClient, ns)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (ph *phaseHandler) reconcileUser(kcUser, specUser *v1alpha1.KeycloakUser, realmName string, authenticatedClient keycloak.KeycloakInterface, ns string) error {
+	// create the user
+	// generate a password and update the users pass
+	// output the result to a secret
+	if kcUser == nil {
+		err := authenticatedClient.CreateUser(specUser, realmName)
+		if err != nil {
+			return err
+		}
+		// generate and update password
+		u, err := authenticatedClient.FindUserByEmail(specUser.UserName, realmName)
+		if err != nil {
+			return errors.Wrap(err, "failed find user")
+		}
+		if u == nil {
+			return errors.New("failed to find user " + specUser.Email)
+		}
+		var newPass string
+		if specUser.Password != nil {
+			newPass = *specUser.Password
+		} else {
+			newPass, err = keycloak.GeneratePassword()
+			if err != nil {
+				return err
+			}
+		}
+		if err := authenticatedClient.UpdatePassword(u, realmName, newPass); err != nil {
+			return errors.Wrap(err, "failed to update password for user "+u.Email)
+		}
+		data := map[string][]byte{"username": []byte(specUser.UserName), "password": []byte(newPass)}
+		userSecret := corev1.Secret{
+			TypeMeta: v1.TypeMeta{
+				APIVersion: "v1",
+				Kind:       "Secret",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Labels:    map[string]string{"application": "sso", "realm": realmName},
+				Namespace: ns,
+				Name:      *specUser.OutputSecret,
+			},
+			Data: data,
+			Type: "Opaque",
+		}
+		if _, err := ph.k8sClient.CoreV1().Secrets(ns).Create(&userSecret); err != nil {
+			return errors.Wrap(err, "failed to create secret ")
+		}
+
+	} else {
```

Comment:
I can't do it that way, I have to check if the kcUser doesn't exist first, as that issues a create. If I checked they are not equal, then if the kcUser is nil, they will not be equal, and it will try to issue an update and not do the create.

e.g.
```
if(!resourcesEqual(kcUser.KeycloakApiUser, specUser.KeycloakApiUser) {
    updateUser()
    return
}
createUser()
```

With a nil kcUser, I'd never get to the `createUser` call